simplify the --interpreter-constraints-whitelist option, and improve performance of python checkstyle

cosmicexplorer opened the pull request:
### Problem

- In the python checkstyle task, we are creating a pex to perform the checking in the task workdir, which goes away after a clean-all.

This affects the checkstyle task's runtime, and since this runs as part of this repo's pre-commit hook, that's significant.

### Solution

- Create the checker pex in `~/.cache/pants/python-checkstyle-checker/` to improve cacheability in non-development cases.
- Make use of the `self.skip_execution` and `self.act_transitively` properties inherited from the checkstyle base class to respect the `--skip` and `--transitive` options which are already registered at the `lint` goal level.

### Result

Python checkstyle is cached a little more often.

cosmicexplorer commented:
This has a one-line change that I've split out into #7013, which should be merged before this one.

cosmicexplorer commented:
Also, I'm going to switch the default for the `--enable-py3-lint` option to `True`, as this makes lots more sense, and I believe the only place we want it to be turned off is in our internal monorepo.

cosmicexplorer commented:
In f59fbeb, I removed the `self._constraints_probably_include_py3()` heuristic in favor of checking if any of the interpreters that `self._interpreter_cache.setup(filters=filters)` returns are python 3 interpreters. I thought that this would more accurately represent what users would actually care about, since if there are no interpreters that can be found which are python 3 interpreters (this is *totally separate* from the default interpreter constraints), the user is going to be completely unable to run their code as python 3, and I therefore doubt the user would then care whether their target's constraints might potentially imply usage in a python 3 context. Also, the `removal_version='1.14.0.dev2'` is fast approaching anyway.

stuhood commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -196,60 +189,99 @@ def checkstyle(self, interpreter, sources):
+  def _interpreter_is_py3(self, interpreter):
+    return interpreter.identity.version[0] == 3
+
+  def _partition_targets_by_min_interpreter(self, tgts):
```
This feels like code we might already have somewhere. Would be good to wait for a John review before landing.

cosmicexplorer replied:
Will do, github's list of my open PRs isn't going to let me forget about it and this is isn't blocking anything time-sensitive so we can easily wait until then.

cosmicexplorer commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -76,14 +77,16 @@ def register_options(cls, register):
+             removal_hint='Python 3 linting will eventually be turned on without a flag',
+             help='Enable linting on Python 3-compatible targets.')
+
+  @memoized_property
```
Done!

cosmicexplorer commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -196,60 +189,99 @@ def checkstyle(self, interpreter, sources):
+      deprecated_conditional(
+        predicate=lambda: py3_was_encountered and not self.get_options().enable_py3_lint,
+        removal_version='1.14.0.dev2',
+        entity_description="This warning",
```
This lets us remove the `TODO` there, thanks!

cosmicexplorer commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -98,28 +101,18 @@ def checker_target(self):
+      pex_base_dir = self.workdir
     else:
       checker_id = hash_all([self._CHECKER_REQ])
+      pex_base_dir = os.path.join(get_pants_cachedir(), 'python-checkstyle-checker')
```
These should have been separated. Will narrow this PR to the first and then make a second one just removing the `--interpreter-constraints-whitelist` option for `1.14.0.dev2` which is coming up extremely soon regardless.

jsirois replied:
I'm confiused. The deprecated conditional was `lambda: self.get_options().interpreter_constraints_whitelist is None` and the deprecation message suggest _using_ `--interpreter-constraints-whitelist`, not getting rid of its use. AFAICT as a user of this beast, `--interpreter-constraints-whitelist` has not yet been deprecated.

cosmicexplorer replied:
Yes -- I hadn't figured that bit out when I left the above comment. I left a [comment to that effect above](https://github.com/pantsbuild/pants/pull/7011#discussion_r245083185) -- if we need to respect the standard deprecation cycle for `--interpreter-constraints-whitelist` (making it e.g. `removal_version=1.16.0.dev1` for that option), we can do that here, there shouldn't be any rush.

cosmicexplorer commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -216,43 +209,90 @@ def _constraints_are_whitelisted(self, constraint_tuple):
 
+  class CheckstyleSetupError(TaskError): pass
+
+  def _parse_requirement(self, filt):
```
This is an artifact of a hasty rebase. I will look over the history again to ensure there are no more such issues. Narrowing the PR to just using the cache dir when not in PANTS_DEV mode will make this easier.

cosmicexplorer replied:
removed!

cosmicexplorer commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -76,14 +77,16 @@ def register_options(cls, register):
-                  'set of compatibility constraints.')
+             help='If disabled, prevent pants from exiting with a failure, but still produce '
+                  'output for style problems.')
+    register('--enable-py3-lint', fingerprint=True, default=True, type=bool,
```
I have reverted this change and added a `removal_version=1.14.0.dev2` to this particular option, which I'm not sure is allowed, but should have been done in the first place -- if we need to respect the deprecation period for this we can extend it.

cosmicexplorer commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -216,43 +209,90 @@ def _constraints_are_whitelisted(self, constraint_tuple):
     return all(version.parse(constraint) in self._acceptable_interpreter_constraints
            for constraint in constraint_tuple)
 
+  class CheckstyleSetupError(TaskError): pass
```
Now used instead of a blanket `TaskError`.

cosmicexplorer commented:
I've removed the totally orthogonal bits about coalescing by minimum interpreter and dumped that into #7020. I've also put back the `--interpreter-constraints-whitelist` option, and added a `removal_version` for that option of `1.14.0.dev2`, as this option is confusing to use and I doubt people have been able to make effective use of it -- but if that's not a good reason, I will change it to `1.16.0.dev0`.

jsirois commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -79,12 +82,17 @@ def register_options(cls, register):
              help='Prevent test failure but still produce output for problems.')
     register('--interpreter-constraints-whitelist', fingerprint=True,
              default=None,
+             removal_version='1.14.0.dev2',
```
If there's no rush as you allude below, it would be good to stick to our documented policy and make this 1.16.0.dev2.

jsirois replied:
What's the hint - as noted it probably makes sense to pull Chris in and make sure the deprecation story is sorted and explained rationally in the user warning messaging.

jsirois commented on src/python/pants/backend/python/tasks/select_interpreter.py:
```diff
@@ -56,10 +56,6 @@ def product_types(cls):
     return [PythonInterpreter]
 
   def execute(self):
-    # NB: Downstream product consumers may need the selected interpreter for use with
```
I think this comment still applies - see line 64 on the RHS. Perhaps the comment is in the wrong location.

jsirois commented on contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py:
```diff
@@ -216,19 +226,34 @@ def _constraints_are_whitelisted(self, constraint_tuple):
+
       for filters, targets in tgts_by_compatibility.items():
         if self.get_options().interpreter_constraints_whitelist is None and not self._constraints_are_whitelisted(filters):
           deprecated_conditional(
```
Can you please pull Chris into this review and have hime help align the deprecated_conditional language below with the now deprecated `--interpreter-constraints-whitelist` option?

Eric-Arellano commented:
Closing because we have deprecated the `python.checks` contrib plugin and will soon be removing it.
